openstack/02a-MAClow/IEEE802154_security.c

```diff
@@ -0,0 +1,861 @@
+\author Giuseppe Piro <giuseppe.piro@poliba.it>,
+\author Gennaro Boggia <gennaro.boggia@poliba.it>,
+\author Luigi Alfredo Grieco <alfredo.grieco@poliba.it>
+\author Malisa Vucinic <malishav@gmail.com>, June 2015.
+*/
+
+#include "packetfunctions.h"
+#include "crypto_engine.h"
+#include "IEEE802154.h"
+#include "IEEE802154E.h"
+#include "idmanager.h"
+#include "openserial.h"
+#include "IEEE802154_security.h"
+
+//=============================define==========================================
+
+//=========================== variables =======================================
+
+ieee802154_security_vars_t ieee802154_security_vars;
+
+//=========================== prototypes ======================================
+
+void IEEE802154_security_getFrameCounter(macFrameCounter_t reference,
+                                        uint8_t*          array);
+uint8_t IEEE802154_security_authLengthChecking(uint8_t securityLevel);
+
+uint8_t IEEE802154_security_auxLengthChecking(uint8_t KeyIdMode,
+                                             bool    frameCounterSuppression,
+                                             uint8_t frameCounterSize);
+
+bool IEEE802154_security_incomingKeyUsagePolicyChecking(m_keyDescriptor* keydesc,
+                                                       uint8_t          frameType,
+                                                       uint8_t          cfi);
+
+bool IEEE802154_security_incomingSecurityLevelChecking(m_securityLevelDescriptor* secLevDesc,
+                                                      uint8_t                    secLevel,
+                                                      bool                       exempt);
+
+m_securityLevelDescriptor* IEEE802154_security_securityLevelDescriptorLookup(uint8_t frameType,
+                                                                            uint8_t cfi);
+
+m_deviceDescriptor* IEEE802154_security_deviceDescriptorLookup(open_addr_t*     address,
+                                                              open_addr_t*     PANID,
+                                                              m_keyDescriptor* keyDescriptor);
+
+m_keyDescriptor* IEEE802154_security_keyDescriptorLookup(uint8_t      KeyIdMode,
+                                                        open_addr_t* keySource,
+                                                        uint8_t      KeyIndex,
+                                                        open_addr_t* DeviceAddress,
+                                                        open_addr_t* panID,
+                                                        uint8_t      frameType);
+
+//=========================== admin ===========================================
+
+/**
+\brief Initialization of security tables and parameters.
+*/
+void IEEE802154_security_init(void){
+
+   //Setting UP Phase
+
+   //MASTER KEY: OpenWSN
+   memset(&ieee802154_security_vars.M_k[0], 0, 16);
+   ieee802154_security_vars.M_k[0] = 0x4e;
```

Comment:
@twatteyne at this moment we have only a single key, used to authenticate beacons and encrypting + authenticating all other frames, i.e., data, ack and command frames. 

In the last version of 6TiSCH minimal (v9) I'm reading that "Depending on security policy, these keys could be the same (i.e., K1=K2)". 

Therefore, can we continue to have a single key or it would be a better choice to implement K2 also as a different data string?